File: android/src/main/java/com/fortysevendeg/mvessel/test/ui/MainActivity.java
package com.fortysevendeg.mvessel.test.ui;

import android.app.Activity;
import android.app.LoaderManager;
import android.content.Context;
import android.content.Loader;
import android.database.Cursor;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CursorAdapter;
import android.widget.ListView;
import android.widget.TextView;
import com.fortysevendeg.mvessel.test.R;
import com.fortysevendeg.mvessel.test.db.ContactsCursorLoader;
import com.fortysevendeg.mvessel.test.db.ContactsOpenHelper;

public class MainActivity extends Activity implements LoaderManager.LoaderCallbacks<Cursor> {

    ContactsOpenHelper contactsOpenHelper;
    ListView listView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        contactsOpenHelper = new ContactsOpenHelper(this);
        contactsOpenHelper.open();

        setContentView(R.layout.sample_main);

        listView = (ListView) findViewById(R.id.list_view);

        getLoaderManager().restartLoader(0, null, this);
    }

    @Override
    protected void onDestroy() {
        contactsOpenHelper.close();
        contactsOpenHelper = null;
        super.onDestroy();
    }

    @Override
    public Loader<Cursor> onCreateLoader(int id, Bundle args) {
        return new ContactsCursorLoader(this, contactsOpenHelper.getDatabase());
    }

    @Override
    public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
        ContactsAdapter adapter = new ContactsAdapter(this, data);
        listView.setAdapter(adapter);
    }

    @Override
    public void onLoaderReset(Loader<Cursor> loader) {

    }

    static class ContactsAdapter extends CursorAdapter {

        private final int nameColumn;
        private final int ageColumn;

        public ContactsAdapter(Context context, Cursor c) {
            super(context, c, false);
            nameColumn = c.getColumnIndex(ContactsOpenHelper.C_NAME);
            ageColumn = c.getColumnIndex(ContactsOpenHelper.C_AGE);
        }

        @Override
        public View newView(Context context, Cursor cursor, ViewGroup parent) {
            LayoutInflater inflater = (LayoutInflater) parent.getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            return inflater.inflate(android.R.layout.simple_list_item_1, parent, false);
        }

        @Override
        public void bindView(View view, Context context, Cursor cursor) {
            TextView textView = (TextView) view.findViewById(android.R.id.text1);
            textView.setText(String.format("%s (%d)", cursor.getString(nameColumn), cursor.getInt(ageColumn)));
        }
    }
}

File: samples/scala-android/src/main/scala/com/fortysevendeg/mvessel/sample/scala/ui/MainActivity.scala
package com.fortysevendeg.mvessel.sample.scala.ui

import android.app.{Activity, LoaderManager}
import android.content.{Context, Loader}
import android.database.Cursor
import android.os.Bundle
import android.view.{LayoutInflater, View, ViewGroup}
import android.widget.{CursorAdapter, ListView, TextView}
import com.fortysevendeg.mvessel._
import com.fortysevendeg.mvessel.sample.scala.db.{ContactsCursorLoader, ContactsOpenHelper}
import com.fortysevendeg.mvessel.sample.scala.{R, TR, TypedFindView}

class MainActivity
  extends Activity with LoaderManager.LoaderCallbacks[Cursor]
  with TypedFindView {

  private[this] lazy val contactsOpenHelper: ContactsOpenHelper = new ContactsOpenHelper(this)
  private[this] lazy val listView: ListView = findView(TR.list_view)

  protected override def onCreate(savedInstanceState: Bundle) = {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.sample_main)
    (contactsOpenHelper, listView)
    getLoaderManager.restartLoader(0, javaNull, this)
  }

  protected override def onDestroy() = {
    contactsOpenHelper.close()
    super.onDestroy()
  }

  def onCreateLoader(id: Int, args: Bundle): Loader[Cursor] = {
    new ContactsCursorLoader(this, contactsOpenHelper.database)
  }

  def onLoadFinished(loader: Loader[Cursor], data: Cursor) = {
    val adapter: ContactsAdapter = new ContactsAdapter(this, data)
    listView.setAdapter(adapter)
  }

  def onLoaderReset(loader: Loader[Cursor]) = {}
}

class ContactsAdapter(context: Context, c: Cursor)
  extends CursorAdapter(context, c, false) {

  val nameColumn: Int = c.getColumnIndex(ContactsOpenHelper.C_NAME)
  val ageColumn: Int = c.getColumnIndex(ContactsOpenHelper.C_AGE)

  def newView(context: Context, cursor: Cursor, parent: ViewGroup): View =
    parent.getContext.getSystemService(Context.LAYOUT_INFLATER_SERVICE) match {
      case inflater: LayoutInflater =>
        inflater.inflate(android.R.layout.simple_list_item_1, parent, false)
      case _ =>
        throw new IllegalStateException("Can't get a layout reference")
    }

  def bindView(view: View, context: Context, cursor: Cursor) {
    val textView: TextView = view.findViewById(android.R.id.text1).asInstanceOf[TextView]
    textView.setText(s"${cursor.getString(nameColumn)} (${cursor.getInt(ageColumn)})")
  }
}